bindings/go/src/fdb/tenant.go

```diff
@@ -0,0 +1,256 @@
+		return errTenantExists
+	}
+
+	t.Set(key, nil)
+
+	return nil
+}
+
+// CreateTenant creates a new tenant in the cluster. The tenant name cannot
+// start with the \xff byte.
+func (d Database) CreateTenant(name KeyConvertible) error {
+	_, err := d.Transact(func(t Transaction) (interface{}, error) {
+		return nil, t.CreateTenant(name)
+	})
+	return err
+}
+
+// DeleteTenant deletes an existing tenant in the cluster.
+func (t Transaction) DeleteTenant(name KeyConvertible) error {
+	if err := t.Options().SetSpecialKeySpaceEnableWrites(); err != nil {
+		return err
+	}
+
+	key := append(Key("\xFF\xFF/management/tenant_map/"), name.FDBKey()...)
+
+	exist, err := t.checkTenantExist(key)
+	if err != nil {
+		return err
+	}
+	if !exist {
+		return errTenantNotFound
+	}
+
+	t.Clear(key)
+
+	return nil
+}
+
+// DeleteTenant creates a new tenant in the cluster. The tenant name cannot
+// start with the \xff byte.
+func (d Database) DeleteTenant(name KeyConvertible) error {
+	_, err := d.Transact(func(t Transaction) (interface{}, error) {
+		return nil, t.DeleteTenant(name)
+	})
+	return err
+}
+
+// ListTenants lists all existings tenants in the cluster.
+func (t Transaction) ListTenants() ([]Key, error) {
+	if err := t.Options().SetSpecialKeySpaceEnableWrites(); err != nil {
+		return nil, err
+	}
+
+	kr, err := PrefixRange(Key("\xFF\xFF/management/tenant_map/"))
+	if err != nil {
+		return nil, err
+	}
+
+	rawTenants, err := t.GetRange(kr, RangeOptions{}).GetSliceWithError()
+	if err != nil {
+		return nil, err
+	}
+
+	// trim tenant prefix
+	tenants := make([]Key, len(rawTenants))
+	for i, rt := range rawTenants {
+		tenants[i] = rt.Key[25:] // len of tenant prefix
+	}
+
+	return tenants, nil
+}
+
+// CreateTenant creates a new tenant in the cluster. The tenant name cannot
+// start with the \xff byte.
+func (d Database) ListTenants() ([]Key, error) {
+	tenants, err := d.Transact(func(t Transaction) (interface{}, error) {
+		return t.ListTenants()
+	})
+	return tenants.([]Key), err
+}
+
+func (t Transaction) checkTenantExist(tenantPath Key) (bool, error) {
+	buf, err := t.Get(tenantPath).Get()
+	if err != nil {
+		return true, err
+	}
+
+	return buf != nil, nil
+}
+
+// Tenant is a handle to a FoundationDB tenant. Tenant is a lightweight
+// object that may be efficiently copied, and is safe for concurrent use by
+// multiple goroutines.
+type Tenant struct {
+	*tenant
+
+	db Database
+}
+
+type tenant struct {
+	ptr *C.FDBTenant
+}
+
+// OpenTenant returns a tenant handle identified by the given name. All transactions
+// created by this tenant will operate on the tenant’s key-space.
+func (d Database) OpenTenant(name KeyConvertible) (Tenant, error) {
+	var outt *C.FDBTenant
+	if err := C.fdb_database_open_tenant(d.database.ptr, byteSliceToPtr(name.FDBKey()), C.int(len(name.FDBKey())), &outt); err != 0 {
+		return Tenant{}, Error{int(err)}
+	}
+
+	tnt := &tenant{outt}
+	runtime.SetFinalizer(tnt, (*tenant).destroy)
+
+	return Tenant{tnt, d}, nil
```

Comment:
It isn't clear to me whether you are talking about efficiency or correctness? Since the tenant object wraps a C object, and has a finalizer I would guess this code is correct. If I stuff the tenant in a map, how is the tenant removed from the map? Does this need to connect with DeleteTenant and such also? Feels like a bit of a can of worms.

Reply:
It was more a question around efficiency. For now it's fine to keep it as it is without making the code more complex. We can add the map later if we see a need for it.